key.core/src/main/java/de/uka/ilkd/key/java/statement/SetStatement.java

```diff
@@ -0,0 +1,76 @@
+/* This file is part of KeY - https://key-project.org
+ * KeY is licensed under the GNU General Public License Version 2
+ * SPDX-License-Identifier: GPL-2.0-only */
+package de.uka.ilkd.key.java.statement;
+
+import de.uka.ilkd.key.java.PositionInfo;
+import de.uka.ilkd.key.java.ProgramElement;
+import de.uka.ilkd.key.java.visitor.Visitor;
+import de.uka.ilkd.key.speclang.njml.JmlParser;
+
+
+
+/**
+ * JML set statement
+ *
+ * @author Julian Wiesler
+ */
+public class SetStatement extends JavaStatement {
+
+    /**
+     * The target of the assignment as a term. Either a heap access for a ghost field
+     * or a ghost variable.
+     */
+    public static int INDEX_TARGET = 0;
+
+    /**
+     * The value of the assignment as a term.
+     */
+    public static int INDEX_VALUE = 1;
+
+    /**
+     * The parser context of the statement produced during parsing.
+     */
+    private final JmlParser.Set_statementContext context;
+
+    /** Constructor used in recoderext */
+    public SetStatement(JmlParser.Set_statementContext context, PositionInfo positionInfo) {
+        super(positionInfo);
+        this.context = context;
+    }
+
+    /** Constructor used when cloning */
+    public SetStatement(SetStatement copyFrom) {
+        this(copyFrom.context, copyFrom.getPositionInfo());
+    }
+
+    /**
+     * Removes the attached parser context from this set statement
+     *
+     * @deprecated weigl: The use of {@link org.antlr.v4.runtime.ParserRuleContext} directly is
+     *             discouraged. It adds
+     *             an (often) unnecessary dependency to the ANTLR across the project.
+     * @return the parser context that was attached
+     */
```

Comment:
What should one use instead? Perhaps mention that it is to be removed with new java+jml framework.

Reply:
I encapsulate the ParserRuleContexts normally in the [`KeyAst`](https://github.com/KeYProject/key/blob/e0995588635b2e5e38c1791e254de76ab4391d36/key.core/src/main/java/de/uka/ilkd/key/nparser/KeyAst.java) classes, which holds the PRC as an attribute but do not expose this. Instead, it offers some basic functionality. 

Of course, you can retrieve a PRC in these classes, but you have to use `ParserFacade#getParseRuleContext` which is the only leakage. 

(Sometimes, in rare situations you need the PRCs, but mostly only in the core framework.)